go/types/list.go

```diff
@@ -170,33 +173,35 @@ func (l List) IterAll(f listIterAllFunc) {
 	}
 }
 
-func (l List) copyReadAhead(out []Value, startIdx uint64) (numBytes uint64) {
-	llen := l.Len()
-	d.PanicIfFalse(startIdx < llen)
+func copyReadAhead(col Collection, blockLength uint64, startIdx uint64) (out []Value, numBytes uint64) {
+	d.PanicIfTrue(blockLength == 0)
 
-	endIdx := startIdx + uint64(len(out))
-	if endIdx > llen {
-		endIdx = llen
-	}
+	l := col.Len()
+	d.PanicIfFalse(startIdx < l)
 
-	if startIdx == endIdx {
-		return
-	}
+	endIdx := startIdx + blockLength
+	d.PanicIfFalse(endIdx <= l)
+
+	valuesPerIdx := getValuesPerIdx(col.asSequence())
+	out = make([]Value, blockLength*valuesPerIdx)
 
-	leaves, localStart := LoadLeafNodes([]Collection{l}, startIdx, endIdx)
+	leaves, localStart := LoadLeafNodes([]Collection{col}, startIdx, endIdx)
 	endIdx = localStart + endIdx - startIdx
 	startIdx = localStart
+	numValues := 0
 
 	for _, leaf := range leaves {
-		ls := leaf.asSequence().(listLeafSequence)
+		seq := leaf.asSequence()
+		values := seq.valuesSlice(startIdx, endIdx)
+		copy(out[numValues:], values)
 
-		values := ls.valuesSlice(startIdx, endIdx)
-		copy(out, values)
-		out = out[len(values):]
-
-		endIdx = endIdx - uint64(len(values)) - startIdx
+		numValues += len(values)
+		endIdx = endIdx - uint64(len(values))/valuesPerIdx - startIdx
 		startIdx = 0
-		numBytes += uint64(len(ls.buff))
+
+		// TODO: this is wrong, it should only include the bytes of |values| only,
```

Comment:
probably not wrong enough to worry about.

Reply:
ok

Reply:
(kept a little note there, for future reference)